### DISCUSSION TITLE
Transformations: "Binary Operator" on timeseries is missing "metric selecting" which is present with table data

### DISCUSSION THREAD

**Discussion by Ultraseamus** [grafana/grafana#38393] [Dashboards, Panels & Visualizations]
When using "Binary Operator" with timeseries data, you have a dropdown that lets you pick from individual timeseries, or add your own constants (which is really great). And when you are working with table data it switches away from showing you each individual field (because those are on different rows, and creating new columns that only apply to single rows would be messy) and instead lets you select from a list of metrics. So you can write one transform that will do math on every row, which is also great.

It does not make much sense to bring over the Timeseries functionality to Table (as I said above, with table data is makes very little sense to run a calculation and generate a new field on just one row). But the table functionality makes perfect sense to be ported over to Timeseries calculations (as an addition, not a replacement). As it is now, with a timeseries there is no way (with transforms) to do something simple like multiply every series by 2 (you could achieve this in other ways, but this is just a simple example to demonstrate).

If more convincing examples of why this would be valuable are needed, I can provide them. Things like doing math between timeseries metrics that were collected from multiple queries. And I think it would also open the door for more powerful Transforms further down the road (things like normalizing each series data).

"Reduce Row" already hints at this same limitation. It is a single calculation that can be applied to multiple series at once, and defaults to having them all selected.

**Comment by ryantxu** [comment]
can you try adding an "Outer join" by time first?

**Comment by Ultraseamus** [comment]
So far as I can tell, that does not change things.

![image](https://user-images.githubusercontent.com/19274602/85882917-fb74cc80-b794-11ea-9286-0752400cd46f.png)

It's not actually an important detail that I was trying to do this math between two separate queries (so long as you join them, but Grafana is already extremely proactive about joining timeseries data). I just made that part of my example because I was worried that someone might shrug off my request by suggesting I use a "Bucket Script". :)

Either way, the "Binary Operator" options still force me to select individual series (which is a great option to have, just not the only one we need). In the feature I am proposing, both of these options would still be here, but each "metric" used in the query would also be listed (in ES, at least, each metric would represent a unique "series of timeseries").

If I selected the metric and multiplied it by two then every series for that metric would have a new X2 line drawn, just as how with a table every row would have a new column created. This is how tables work today. It would be the same as if I selected each individual series and multiplied by 2, but there could be 50 series with shifting names, and it would still work.

Trying to implement this feature on my own, I can see why it is not working. It is because timeseries and table data are still distinct from each other. The timeseries is passed in as a mutabledataframe, table data is not. Ironically... I think I would actually have a much easier time implementing this feature if Grafana were not so proactive at automatically joining timeseries data. No matter what I do, those series always get passed into the "CalculateField" "initOptions" as one big series all mashed together. I would argue that this only happen if someone actually applied a "Outer Join" on Time. As it is now, it takes away so much control from the user.

**Comment by torkelo** [comment]
Not sure I understand the issue, is it that you cannot use the Join as every series has different time index/points so the join is causing issues?

**Comment by torkelo** [comment]
When I use the binary operation I can select series

**Comment by Ultraseamus** [comment]
I feel the word "series" is a bit overloaded here, so I am having a hard time explaining the issue. Also, I am trying to avoid using terminology that is specific to Elasticsearch. In the query I am writing, performing a join on Time makes no difference, because it is automatically joined on Time. What do I call a series of series? The result of a Terms agg + Date agg + Metric agg.

Basically, when using "Timeseries" data, every "series" (each one mapping to a single line on the graph) can be selected individually. I can add line one to line two, and create a line three.

But if I have line (Series) one, two, three... let's say all the way to 50, created from a single query (Terms agg + Date agg + Metric agg). There's no way that I can see to say that I want every one of those lines (series) to be multiplied by 2. Achieving that with transforms on timeseries data would be almost impossible, I'd have to select each series individually, and it would fall apart if the series names changed (like they would if the terms agg was sorted by some realtime metric).

This is distinctly different from "Table Data" where each "Series" essentially maps to a single row. Rather than having to reference each row individually to perform math, I can select the "metric" belonging to that "Series" (which is common to every "series" in that table). With a single selection I can generate a new column for every "Series" (row).

**Comment by ryantxu** [comment]
As implemented -- the binary operator works on values within the same row of a DataFrame.  If the input is all simple TimeSeries (multiple DataFrames each with a time/value column) this transformer will first apply a "join by time" transformer so that the multiple value columns all exist in the same row.

I still not get what you are asking to change.

**Comment by Ultraseamus** [comment]
In the image I shared above, you can see that the "Series" are being joined by "Time". And in the dropdown I can select either Series 1 or Series 2.

Here is what it looks like if I remove the "Date Histogram" and turn it into Table Data:

![image](https://user-images.githubusercontent.com/19274602/86062688-1efe7800-ba1e-11ea-9c9a-59d995ae3f1e.png)

Instead of selecting "Series 1" or "Series 2" (like with timeseries data) I am selecting the "Metric". If I added back the Date histogram, my options would no longer include "Sum" it would turn into "Series 1" and "Series 2".

Sorry that I am doing a very poor job at explaining this... but I want to be able to add back the date Histogram and have my options be: "Series 1, Series 2, Sum". If I selected Sum, every line would be affected by the operation (it would run against every series generated by that Metric), and the number of lines would double. If I selected just "Series 1 or Series 2" then only a single new line would be created per Binary Operation.

In the table example a row equates to a line in the graph example. But they both work differently. For table data the binary operator generates a new column for every row, it does not generate a new row. Unlike the Graph where each binary operator generates a new line.

> If the input is all simple TimeSeries (multiple DataFrames each with a time/value column) this transformer will first apply a "join by time" transformer so that the multiple value columns all exist in the same row.

I am dealing specifically with a slightly more complex TimeSeries. Mine have Time + Field + Metirc (the existence of a field is important, I think). Each field is a separate line when it is a Timeseries. And a separate row when it is Table data. And we don't necessarily want each of these series fields to exist on the same row (if that were happening with "Table Data" it would be very restricting), we want them to each have their own rows, with a shared "Metric" column. So that we can have a binary operator that essentially creates a new "metric" column for every row (iterating over each TimeSeries).

**Comment by edin008** [comment]
Any update on this issue? Would be nice to have this feature, to have transformation with possibility to apply one binary operation across all metrics on graph. In my case I cannot make binary operation in query because Im using OpenTSDB.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #83011. Feel free to continue the discussion around this there. Thank you!
